src/hash_builder/mod.rs

```diff
@@ -201,20 +216,27 @@ impl HashBuilder {
         let mut build_extensions = false;
         // current / self.key is always the latest added element in the trie
         let mut current = self.key.clone();
+
         debug_assert!(!current.is_empty());
 
         trace!(target: "trie::hash_builder", ?current, ?succeeding, "updating merkle tree");
 
         let mut i = 0usize;
+        let trie_default = &TrieMask::default();
         loop {
             let _span = tracing::trace_span!(target: "trie::hash_builder", "loop", i, ?current, build_extensions).entered();
 
+            let state_len = self.state_masks.len();
+            let current_len = current.len();
+            let succeding_empty = !succeeding.is_empty();
+
             let preceding_exists = !self.state_masks.is_empty();
-            let preceding_len = self.state_masks.len().saturating_sub(1);
+            let preceding_len = state_len.saturating_sub(1);
+            let s_and_p = succeding_empty || preceding_exists;
```

Comment:
this greatly reduces readability, so we should do it only if we're sure that compiler will not optimize it on its own. I think the compiler will take care of it for us.